lib/active_support/cache/redis_store.rb

```diff
@@ -149,6 +149,45 @@ def fetch_multi(*names)
         fetched
       end
 
+      # Performs a CAS operation (like MemcachedStore) using a watch,
+      # ensuring an optimistic lock on the given name
+      def cas(name, options = nil)
+        options = merged_options(options)
+        key = normalize_key(name, options)
+        result = false
+        instrument(:cas, name, options) do
+          with do |c|
+            c.watch(key) do
+              entry = read_entry(key, options)
+              if entry
+                new_entry = yield entry.value
+                result = c.multi { write key, new_entry, options }
+              else
+                result = false
```

Comment:
I believe you need to either `unwatch` or do an empty `multi` here, otherwise the watched state will leak from this method.
Just confirmed with `redis-cli monitor`.